### PR TITLE
feat: juju

### PR DESCRIPTION
**Pull request by yanksyoon**
This requires a force merge because there's a bug with Juju testing w/ cleanup of the tests.

https://github.com/canonical/github-runner-image-builder-operator/actions/runs/11359657961/job/31601569520#step:23:2140

All tests are successful but the cleanup fails.

Applicable spec: <link>

### Overview

Introduces Juju installed & bootstrapped images.

### Rationale

- To have images w/ Juju installed & bootstrapped which reduces the operator workflow time by a large margin.

### Juju Events Changes

None.

### Module Changes

- `builder.py` now calls image builder w/ Juju option.
- `state.py` parses juju config option.

### Checklist

- [x] The [charm style guide](https://juju.is/docs/sdk/styleguide) was applied
- [x] The [contributing guide](https://github.com/canonical/is-charms-contributing-guide) was applied
- [x] The changes are compliant with [ISD054 - Managing Charm Complexity](https://discourse.charmhub.io/t/specification-isd014-managing-charm-complexity/11619)
- [x] The documentation is generated using `src-docs`
- [x] The documentation for charmhub is updated.
- [x] The PR is tagged with appropriate label (`urgent`, `trivial`, `complex`)

### CODE REVIEW

File: tests/unit/test_charm.py [hunk 1]
```diff
@@ -17,12 +16,14 @@
...
 from charm import GithubRunnerImageBuilderCharm
+from tests.unit import factories
 
 
 @pytest.fixture(name="patch_builder_init_config_from_charm", scope="function")
 def patch_builder_init_config_from_charm(monkeypatch: pytest.MonkeyPatch):
     """Fixture to patch builder init config."""
-    monkeypatch.setattr(
+    # This is to monkeypatch BuilderInitConfig.from_charm function for unit tests.
+    monkeypatch.setattr(  # pylint: disable=expression-not-assigned
```
Comment (jdkandersson) [inline tests/unit/test_charm.py:26]: let's remove this

File: src/charm.py
Comment (yanksyoon) [inline src/charm.py:1]: Refactoring changes.

File: src/image.py [hunk 2]
```diff
@@ -55,7 +55,7 @@ def _on_image_relation_joined(self, event: ops.RelationJoinedEvent) -> None:
             event: The event emitted when a relation is joined.
```
Comment (yanksyoon) [inline src/image.py:55]: Refactoring changes

File: src/state.py [hunk 3]
```diff
@@ -148,24 +153,29 @@ def from_charm(cls, charm: ops.CharmBase) -> tuple["BaseImage", ...]:
...
+                    cls(LTS_IMAGE_VERSION_TAG_MAP[image_name])
+                    if image_name in LTS_IMAGE_VERSION_TAG_MAP
+                    else cls(image_name)
+                )
+                for image_name in image_names
             )
-            for image_name in image_names
-        )
+        except ValueError as exc:
+            raise InvalidBaseImageError(f"Invalid bases: {image_names}") from exc
```
Comment (yanksyoon) [inline src/state.py:178]: Refactoring changes to invert CharmInvalidConfigException dependency

File: tests/integration/conftest.py [hunk 4]
```diff
@@ -167,8 +167,9 @@ async def test_charm_fixture(
 
     yield app
 
-    logger.info("Cleaning up test charm.")
-    await model.remove_application(app_name=app_name)
+    # logger.info("Cleaning up test charm.")
+    # await model.remove_application(app_name=app_name, force=True, no_wait=True)
+    # logger.info("Test charm removed.")
```
Comment (jdkandersson) [inline tests/integration/conftest.py:172]: In a follow up PR could we please remove this commented out code?

I'm not suprised this is flaky, this has never been reliable from memory

File: src/state.py [hunk 5]
```diff
@@ -329,30 +339,64 @@ class BuildConfigInvalidError(CharmConfigInvalidError):
 
 
 @dataclasses.dataclass
-class BuilderRunConfig:
```
Comment (yanksyoon) [inline src/state.py:332]: Refactoring changes due to too-many-attributes

File: src/state.py [hunk 6]
```diff
@@ -362,104 +406,120 @@ def cloud_name(self) -> str:
...
-                ExternalBuildConfig.from_charm(charm=charm) if external_build_enabled else None
-            ),
-            num_revisions=revision_history_limit,
-            prefix=charm.app.name,
-            runner_version=runner_version,
+            image_config=image_config,
         )
 
 
+class BuildIntervalConfigError(CharmConfigInvalidError):
```
Comment (yanksyoon) [inline src/state.py:470]: Exception dependency inversion changes.

**Review by jdkandersson — APPROVED** [review]

File: tests/integration/conftest.py
Comment (yanksyoon) [inline tests/integration/conftest.py:1]: Refactoring changes & test setup for bare image and juju image

File: tests/integration/helpers.py
Comment (yanksyoon) [inline tests/integration/helpers.py:1]: Refactoring changes

File: tests/integration/conftest.py (hunk 4 above)
Comment (yanksyoon) [reply tests/integration/conftest.py:172]: Sure I'll replace the comments with something in the lines of - "due to the bug the cleanup must be done manually on OpenStack"

**Pull request merged**